atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepableCellsReader.java

```diff
@@ -0,0 +1,139 @@
+    private static final TableReference TABLE_REF = TargetedSweepTableFactory.of()
+            .getSweepableCellsTable(null).getTableRef();
+    private static final ColumnRangeSelection ALL_COLUMNS = allPossibleColumns();
+
+    private final KeyValueService kvs;
+
+    SweepableCellsReader(KeyValueService kvs) {
+        this.kvs = kvs;
+    }
+
+    List<WriteInfo> getLatestWrites(long partitionFine, ShardAndStrategy shardAndStrategy) {
+        TargetedSweepMetadata metadata = getDefaultMetadata(shardAndStrategy);
+        SweepableCellsTable.SweepableCellsRow row = SweepableCellsTable.SweepableCellsRow.of(
+                partitionFine, metadata.persistToBytes());
+
+        RowColumnRangeIterator resultIterator = getAllColumns(row);
+
+        Map<TableReferenceAndCell, Long> results = new HashMap<>();
+        resultIterator.forEachRemaining(entry -> populateResults(row, entry, results));
+
+        return results.entrySet().stream()
+                .map(entry -> WriteInfo.of(entry.getKey(), entry.getValue()))
+                .collect(Collectors.toList());
+    }
+
+    private TargetedSweepMetadata getDefaultMetadata(ShardAndStrategy shardAndStrategy) {
+        return ImmutableTargetedSweepMetadata.builder()
+                .conservative(shardAndStrategy.isConservative())
+                .dedicatedRow(false)
+                .shard(shardAndStrategy.shard())
+                .dedicatedRowNumber(0)
+                .build();
+    }
+
+    private RowColumnRangeIterator getAllColumns(SweepableCellsTable.SweepableCellsRow row) {
+        return getAllColumns(ImmutableList.of(row.persistToBytes()));
+    }
+
+    private RowColumnRangeIterator getAllColumns(Iterable<byte[]> rows) {
+        return kvs.getRowsColumnRange(TABLE_REF, rows, ALL_COLUMNS, 1000, SweepQueueUtils.CAS_TS);
+    }
+
+    private void populateResults(SweepableCellsTable.SweepableCellsRow row, Map.Entry<Cell, Value> entry,
+            Map<TableReferenceAndCell, Long> results) {
+        SweepableCellsTable.SweepableCellsColumn col = SweepableCellsTable.SweepableCellsColumn.BYTES_HYDRATOR
+                .hydrateFromBytes(entry.getKey().getColumnName());
+
+        if (isReferenceToDedicatedRows(col)) {
+            populateFromDedicatedRows(row, col, results);
+        } else {
+            populateFromValue(row, col, entry.getValue(), results);
+        }
+    }
+
+    private boolean isReferenceToDedicatedRows(SweepableCellsTable.SweepableCellsColumn col) {
+        return col.getWriteIndex() < 0;
+    }
+
+    private void populateFromDedicatedRows(SweepableCellsTable.SweepableCellsRow row,
+            SweepableCellsTable.SweepableCellsColumn col, Map<TableReferenceAndCell, Long> results) {
+        TargetedSweepMetadata metadata = TargetedSweepMetadata.BYTES_HYDRATOR.hydrateFromBytes(row.getMetadata());
+        long partitionFine = row.getTimestampPartition();
+        int numberOfDedicatedRows = (int) -col.getWriteIndex();
```

Comment:
It occurs to me that we'll have similar logic in reverse for when we're writing to dedicated rows. That logic would ideally live in the same class/layer. This class might then end up doing little more than delegating to one of two readers.

Reply:
You wanted a test suggestion: we should have tests to verify that when you write something to the queue and then read it back, you get the same thing you wrote. That's more important than the above suggestion.